components/ai_chat/core/browser/ai_chat_service.h

```diff
@@ -323,6 +330,9 @@ class AIChatService : public KeyedService,
   // Engine for tab organization, created on demand and owned by AIChatService.
   std::unique_ptr<ai_chat::EngineConsumer> tab_organization_engine_;
 
+  // Tools that are available and shared across all conversations.
+  std::vector<std::unique_ptr<Tool>> tools_;
```

Comment:
I don't see a reason to store a vector of generic Tool instances here. How about just storing a `std::unique_ptr<MemoryTool>` here and avoid the finding in a vector when we want to interact with the specific tools, e.g. in `OnMemoryEnabledChanged`. `GetTools()` is the only method that needs to return a `vector<WeakPtr<Tool>>` and it can just return `{memory_tool_.GetWeakPtr()}`